gc/gc-iceberg/src/main/java/org/projectnessie/gc/iceberg/ExpireSnapshotsProcedure.java

```diff
@@ -0,0 +1,435 @@
+      createGCBranch(expiryBranchName, api);
+      // get Nessie catalog
+      Catalog nessieCatalog = GCUtil.loadNessieCatalog(spark(), gcCatalogName, expiryBranchName);
+      String finalRunId = runId;
+      expiredSnapshotsPerContentId
+          .entrySet()
+          .forEach(
+              entry -> {
+                // Expiry Algorithm:
+                // A. For each content id, fetch ANY one of the content.
+                // B. Use the newly created expiry branch. On this branch,
+                // C. Do a single PUT operation with a fetched content from step-A and
+                //    a dummy ContentKey.
+                // D. Collect the list of all the expired snapshot for this content id.
+                // E. Fetch this table using dummy ContentKey on expiry branch. For this table,
+                // F. call expire_snapshot action from Iceberg by passing the list of expired
+                //    snapshots.
+                // G. Once all the content ids are handled, drop this branch.
+                //    (Else it may impact the next GC operations as some contents are still referred
+                // in
+                //    this branch)
+                String contentId = entry.getKey();
+                addOneContentToExpireProcedureBranch(
+                    expiryBranchName, identifiedResultsRepo, api, finalRunId, contentId);
+                Table table =
+                    nessieCatalog.loadTable(
+                        TableIdentifier.parse(EXPIRE_NAMESPACE + "." + contentId));
+
+                // TODO: Remove after https://github.com/apache/iceberg/pull/4509
+                updateMetadataLocation((BaseTable) table);
+
+                Set<Snapshot> snapshotSet =
+                    new TreeSet<>((o1, o2) -> (int) (o1.timestampMillis() - o2.timestampMillis()));
+                Set<String> processedMetadataFiles = new HashSet<>();
+                List<String> metadataLocations =
+                    new ArrayList<>(metadataLocationsPerContentId.get(contentId));
+
+                metadataLocations.sort(Collections.reverseOrder());
+
+                metadataLocations.forEach(
+                    metadataLocation -> {
+                      if (!processedMetadataFiles.contains(metadataLocation)) {
+                        TableMetadata metadata =
+                            TableMetadataParser.read(table.io(), metadataLocation);
+                        snapshotSet.addAll(metadata.snapshots());
+                        List<String> files =
+                            metadata.previousFiles().stream()
+                                .map(TableMetadata.MetadataLogEntry::file)
+                                .collect(Collectors.toList());
+                        processedMetadataFiles.addAll(files);
+                      }
+                    });
+
+                // create a dummy table metadata which has all the snapshots for a content from all
```

Comment:
If two snapshots from different references have the same snapshot-ID, your approach will not work.

Reply:
> If two snapshots from different references have the same snapshot-ID, your approach will not work.

The probability of this is same as two commits in Iceberg having a same snapshot id allocated. 

snapshot id is UUID: 

https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/TableOperations.java#L116-L120

Reply:
Iceberg does not allow duplicate snapshot IDs - and GC should be resilient against duplicate snapshot IDs in different branches.

Reply:
Sorry, it is a same code for getting snapshot Id as mentioned above. 
As it is UUID, Iceberg believes it as random. 

The same code will be used for branches also. So, it will be unique. 

How can two branches can have same snapshot id for two different commits as it uses the same code?  (ignoring schema change as it doesn't affect gc)